aws/resource_aws_appsync_datasource.go

```diff
@@ -0,0 +1,327 @@
+package aws
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appsync"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsAppsyncDatasource() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsAppsyncDatasourceCreate,
+		Read:   resourceAwsAppsyncDatasourceRead,
+		Update: resourceAwsAppsyncDatasourceUpdate,
+		Delete: resourceAwsAppsyncDatasourceDelete,
+
+		Schema: map[string]*schema.Schema{
+			"api_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					if !regexp.MustCompile(`[_A-Za-z][_0-9A-Za-z]*`).MatchString(value) {
+						errors = append(errors, fmt.Errorf("%q must match [_A-Za-z][_0-9A-Za-z]*", k))
+					}
+					return
+				},
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := strings.ToUpper(v.(string))
+					validTypes := []string{"AWS_LAMBDA", "AMAZON_DYNAMODB", "AMAZON_ELASTICSEARCH"}
+					for _, str := range validTypes {
+						if value == str {
+							return
+						}
+					}
+					errors = append(errors, fmt.Errorf("expected %s to be one of %v, got %s", k, validTypes, value))
+					return
+				},
+				StateFunc: func(v interface{}) string {
+					return strings.ToUpper(v.(string))
+				},
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"dynamodb_config": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"region": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"table": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"use_caller_credentials": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+					},
+				},
+				ConflictsWith: []string{"elasticsearch_config", "lambda_config"},
+			},
+			"elasticsearch_config": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"region": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"endpoint": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+				ConflictsWith: []string{"dynamodb_config", "lambda_config"},
+			},
+			"lambda_config": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"arn": {
```

Comment:
I think that we should make this name a bit more specific, e.g. `function_arn`, so it's obvious that we expect ARN of the Lambda function - just in case Amazon decides to introduce more ARN-based arguments in the future.